web/src/main/java/se/inera/intyg/webcert/web/service/facade/impl/certificatefunctions/CertificateSignAndSendFunctionImpl.java

```diff
@@ -0,0 +1,119 @@
+package se.inera.intyg.webcert.web.service.facade.impl.certificatefunctions;
+
+import static se.inera.intyg.common.support.facade.model.CertificateRelationType.COMPLEMENTED;
+
+import java.util.Optional;
+import org.springframework.stereotype.Component;
+import se.inera.intyg.common.af00213.support.Af00213EntryPoint;
+import se.inera.intyg.common.db.support.DbModuleEntryPoint;
+import se.inera.intyg.common.doi.support.DoiModuleEntryPoint;
+import se.inera.intyg.common.fk7263.support.Fk7263EntryPoint;
+import se.inera.intyg.common.lisjp.support.LisjpEntryPoint;
+import se.inera.intyg.common.luae_fs.support.LuaefsEntryPoint;
+import se.inera.intyg.common.luae_na.support.LuaenaEntryPoint;
+import se.inera.intyg.common.luse.support.LuseEntryPoint;
+import se.inera.intyg.common.support.facade.model.Certificate;
+import se.inera.intyg.common.ts_bas.support.TsBasEntryPoint;
+import se.inera.intyg.common.ts_diabetes.support.TsDiabetesEntryPoint;
+import se.inera.intyg.infra.security.authorities.AuthoritiesHelper;
+import se.inera.intyg.infra.security.common.model.AuthoritiesConstants;
+import se.inera.intyg.webcert.web.web.controller.facade.dto.ResourceLinkDTO;
+import se.inera.intyg.webcert.web.web.controller.facade.dto.ResourceLinkTypeDTO;
+
+@Component
+public class CertificateSignAndSendFunctionImpl implements CertificateSignAndSendFunction {
+
+    private final AuthoritiesHelper authoritiesHelper;
+    private static final String SIGN_AND_SEND_NAME = "Signera och skicka";
+    private static final String SIGN_NAME = "Signera intyget";
+    private static final String SIGN_DESCRIPTION = "Intyget signeras.";
+    private static final String SIGN_AND_SEND_DESCRIPTION_ARBETSFORMEDLINGEN = "Intyget skickas direkt till Arbetsförmedlingen.";
+    private static final String SIGN_AND_SEND_DESCRIPTION_FORSAKRINGSKASSAN = "Intyget skickas direkt till Försäkringskassan.";
+    private static final String SIGN_AND_SEND_DESCRIPTION_SKATTEVERKET = "Intyget skickas direkt till Skatteverket.";
+    private static final String SIGN_AND_SEND_DESCRIPTION_SOCIALSTYRELSEN = "Intyget skickas direkt till Socialstyrelsen.";
+    private static final String SIGN_AND_SEND_DESCRIPTION_TRANSPORTSTYRELSEN = "Intyget skickas direkt till Transportstyrelsen.";
+
+
+    public CertificateSignAndSendFunctionImpl(AuthoritiesHelper authoritiesHelper) {
+        this.authoritiesHelper = authoritiesHelper;
+    }
+
+    @Override
+    public Optional<ResourceLinkDTO> get(Certificate certificate) {
+
+        if (isSignedAndSendDirectly(certificate)) {
```

Comment:
Should we add a check that if the certificate doesn't have the status UNSIGNED return Optional.empty()? Since only drafts can be signed.

Reply:
This check is already made in GetCertificatesAvailableFunctionsImpl so doing this is redundant:

```java

 switch (certificate.getMetadata().getStatus()) {
            case UNSIGNED:
                availableFunctions.addAll(
                    getAvailableFunctionsForDraft(certificate)
                );
                break;
            case SIGNED:
                availableFunctions.addAll(
                    getAvailableFunctionsForCertificate(certificate)
                );
                break;
            case LOCKED:
                availableFunctions.addAll(
                    getAvailableFunctionsForLockedDraft(certificate)
                );
                break;
            case REVOKED:
                availableFunctions.addAll(
                    getAvailableFunctionsForRevokedCertificate(certificate)
                );
                break;
            default:
        }
```

Reply:
The check in AvailableFunctions should exist but now that the code is being refactored it's nice when the logic that is extracted works by itself without logic from the class using it. SrsFunction and CopyCertificateFunction are examples of where the status check is included in the function also. But it works like this too.